### DISCUSSION TITLE
When does juicefs delete the file on s3?

### DISCUSSION THREAD

**Discussion by 201341** [juicedata/juicefs#717] [Q&A] [answered]
I created 100m file, then delete it. But juicefs didn't delete it on s3. So I want to know when does juicefs delete file on s3? By the way, when the file is small, such as 10byte,  jucefs deletes it right away.

**Comment by davies** [comment, marked as answer]
JuiceFS delete the objects in background under limited concurrency, it will retry the unfinished or failed deletion later.

Also, you can use `juicefs gc` to cleanup any leak or delayed deletion.

**Reply by 201341** [reply]
I have already tried `the juicefs gc`, but it doesn't find the leak or delayed deletion. And It did scan the deleted file inode, But the object on s3 didn't be deleted. 
<img width="886" alt="截屏2021-08-09 10 09 19" src="https://user-images.githubusercontent.com/13047679/128654435-ba9caa8e-7465-489f-b757-3de7b15a9c6e.png">
<img width="893" alt="截屏2021-08-09 10 27 12" src="https://user-images.githubusercontent.com/13047679/128654741-90c63a52-fe10-4733-8cbc-8104dc70fcd0.png">

**Reply by davies** [reply]
you need to add `--delete` to delete the objects.

**Reply by 201341** [reply]
I used `./juicefs gc --delete`, it didn't work.

**Reply by xiaogaozi** [reply]
Which version of JuiceFS you use? And please describe how to reproduce this problem.

**Reply by 201341** [reply]
juicefs version
`juicefs version 0.16-dev (2021-08-08 fb280776)`
sqlite3 version
`SQLite version 3.32.3 `
minio version
`2021-04-18T19:26:29Z`

Produce file
`dd if=/dev/zero of=./test4 bs=1024 count=102400`
Adjust
`vim test4` and insert some character.
Remove file
`rm -rf test4`
Gc
`./juicefs gc --delete metaurl`
Observe the minio object

**Reply by 201341** [reply]
Hi, I have know the reason, It's a timezone bug. minio uses UTC, but the juicefs uses time.Now(), which return the current local time.

**Reply by davies** [reply]
Cool, can you create an issue or send a PR?

**Reply by Xuanwo** [reply]
I created an issue about this problem: https://github.com/juicedata/juicefs/issues/1002

However, I don't think it's timezone-related problem: both `time.Now()` and `time.After()` are already taken timezone into consideration. So the different timezone between `juciefs` and `minio` should not affect the gc.

For reference:

https://github.com/juicedata/juicefs/blob/ccb13303eabbed684e392f018a025a684a197012/cmd/gc.go#L220-L231

**Reply by Xuanwo** [reply]
BTW, here is a hardcoded logic to skip all objects that created in 1-hour:

https://github.com/juicedata/juicefs/blob/ccb13303eabbed684e392f018a025a684a197012/cmd/gc.go#L196

Maybe this is the root cause?

**Reply by Xuanwo** [reply]
Already fixed in https://github.com/juicedata/juicefs/pull/728
